### ISSUE TITLE
Stan C++ wrapping code is written in Cython, not pybind11

### ISSUE THREAD

**Issue by riddell-stan** [stan-dev/httpstan#403]
All Cython code should be removed and replaced with C++ code using pybind11. This would eliminate Cython as one of the languages required to understand code in this repository.

The code which needs to be replaced is in the file `httpstan/anonymous_stan_model_services.pyx.template`.

**Comment by ahartikainen** [comment]
Some notes

We would need to replace `httpstan/anonymous_stan_model_services.pyx.template` with `httpstan/anonymous_stan_model_services.cpp.template`

We need to include the model `.hpp` file

    #include <${module_name}.hpp>

We access the namespace with

    namespace py = pybind11;
    using namespace ${model_name}_namespace;

and then use types as 

    py::dict data

We can import python modules as

    py::object httpstan_utils = py::module::import("httpstan").attr("utils");

and then access them like

    py::tuple sdata = httpstan_utils.attr("_split_data")(data);


Implementing a `model` and `var_context` classes are a bit out of my C++ knowledge. We apparently need to implement same functions as in cython solution. 

Do we want to keep the same structure as we have with Cython, where we have many files and then we just include them.

In compilation we need two template arguments

```
def services_extension_module_cpp_code(module_name: str, model_name: str) -> str:
    """Return Cython code wrapping model-specific stan::services functions.

    Arguments:
        module_name: C++ module name.
        model_name: C++ model name..

    Returns:
        str: C++ wrapping code.
    """

    cpp_code_template = importlib.resources.read_text(__package__, "anonymous_stan_model_services.cpp.template")
    stan_model_name = f"model_{model_name.split('/')[1]}"
    return string.Template(cpp_code_template).substitute(module_name=module_name, model_name=stan_model_name)
```

We need to handle both `hpp` and `cpp` files

    cpp_code_path = hpp_code_path.with_suffix(".cpp")

Then at the compilation step (C++) we need to include

```
include_dirs = [
        pybind11.get_include(),
        httpstan_dir,  # for socket_writer.hpp and socket_logger.hpp
        model_directory_path.as_posix(),
        os.path.join(httpstan_dir, "include"),
        os.path.join(httpstan_dir, "include", "lib", "eigen_3.3.7"),
        os.path.join(httpstan_dir, "include", "lib", "boost_1.72.0"),
        os.path.join(httpstan_dir, "include", "lib", "sundials_5.2.0", "include"),
        os.path.join(httpstan_dir, "include", "lib", "tbb_2019_U8", "include"),
    ]
```
For extension we need to add cpp file (I'm not 100% sure about this step)

extension = setuptools.Extension(
        module_name,
        language="c++",
        sources=[cpp_code_path.as_posix(), callbacks_writer_pb_path.as_posix()],
        define_macros=stan_macros,
        include_dirs=include_dirs,
        library_dirs=[f"{PACKAGE_DIR / 'lib'}"],
        libraries=libraries,
        extra_compile_args=extra_compile_args,
        extra_link_args=[f"-Wl,-rpath,{PACKAGE_DIR / 'lib'}"],
    )


TL:DR:  There are couple of places we need to change the code, but in total, it is not much.

I still don't follow different `var_context` handling and the different parts needed for `model / model_base` class. Also pointer/reference handling is not my strong point.

**Comment by riddell-stan** [comment]
I should have mentioned I'm working on this. I'm making good progress. 
Cython will be gone within a week!

**Comment by ahartikainen** [comment]
Great, that is good news (👍). I was studying how to use pybind11, so these notes are from that exercise.
